### PR TITLE
Added support of tab completion via signature

### PR DESCRIPTION
**Pull request by jlstevens**
Work in progress PR to address #405.

### CODE REVIEW

**Comment by jlstevens** [comment]
Tab completion based on ``__signature__`` seems to be working starting at 5c46947. Not quite sure if this PR is ready yet: I had to do a fair bit of filtering on the arguments due to name clashes. I'll see if the tests pass on travis before taking another look (and probably adding a unit test or two).

File: hvplot/__init__.py [hunk 1]
```diff
@@ -55,21 +55,36 @@ def _get_doc(cls, kind, completions=False, docstring=True, generic=True, style=T
...
+        extra_kwargs = _hv.core.util.unique_iterator(valid_opts + kind_opts
+                                                     + converter._axis_options
+                                                     + converter._op_options)
+        filtered_signature = [p for p in sig.parameters.values()
+                              if p.kind != inspect.Parameter.VAR_KEYWORD]
+        extra_params = [inspect.Parameter(k, inspect.Parameter.KEYWORD_ONLY)
+                        for k in extra_kwargs
+                        if k not in [p.name for p in filtered_signature]]
+        all_params = (filtered_signature + extra_params
+                      + [inspect.Parameter('kwargs', inspect.Parameter.VAR_KEYWORD)])
```
Comment (jlstevens) [inline hvplot/__init__.py]: Note that the ``**kwargs`` `VAR_KEYWORD`` at the end here is not strictly needed for tab-completion to work. If we don't except any other ``kwargs`` I am happy to remove it.

File: hvplot/__init__.py [hunk 2]
```diff
@@ -55,21 +55,36 @@ def _get_doc(cls, kind, completions=False, docstring=True, generic=True, style=T
...
+        signature = inspect.Signature(all_params)
+
     parameters += [(o, None) for o in
                    valid_opts+kind_opts+converter._axis_options+converter._op_options]
```
Comment (philippjfr) [inline hvplot/__init__.py]: Should the extra_kwargs processing and this not be merged somehow? Seems like they would be duplicated here as well.

File: hvplot/__init__.py [hunk 3]
```diff
@@ -55,21 +55,36 @@ def _get_doc(cls, kind, completions=False, docstring=True, generic=True, style=T
...
+        signature = None
     else:
         sig = inspect.signature(method)
         for name, p in list(sig.parameters.items())[1:]:
             if p.kind == 1:
                 parameters.append((name, p.default))
+
+        extra_kwargs = _hv.core.util.unique_iterator(valid_opts + kind_opts
+                                                     + converter._axis_options
+                                                     + converter._op_options)
```
Comment (philippjfr) [inline hvplot/__init__.py]: Don't have to change this now but I've adopted this indentation style (which conforms to blacks line wrapping rules) for Panel and hvPlot now:

```python        
extra_kwargs = _hv.core.util.unique_iterator(
    valid_opts + kind_opts + converter._axis_options + converter._op_options
)
```

File: hvplot/__init__.py (hunk 2 above)
Comment (jlstevens) [reply hvplot/__init__.py]: I assume you are talking about `parameters` in which case yes...there will be duplication in the docstring completion hint. The difference is that duplication in the docstring hint doesn't generate an error (and tab completion works fine) while the signature approach validates against duplication.

I can use `unique_iterator` here but I'm not sure it is worth doing anything much more complicated: we will be dropping Python 2 before long (so we won't be using the docstring hint anymore) and the only improvement would be a slight reduction in the verbosity of part of the docstring no one looks at anyway...

Comment (jlstevens) [reply hvplot/__init__.py]: Addressed in bbf6f61.

File: hvplot/__init__.py (hunk 3 above)
Comment (jlstevens) [reply hvplot/__init__.py]: Addressed in bbf6f61.

**Comment by philippjfr** [comment]
Thanks, looks good.

**Pull request merged**